### PR TITLE
Use file handles directly instead of C file descriptors on Win32

### LINKED ISSUE #14389: Introduce platform-specific `FileDescriptor::Handle`
In #10706 we extracted all system-specifics out of the generic `Socket` API. Part of that was the introduction of `Socket::Handle` which is an abstract type for representing a socket handle to the respective operating system.

We should do the same thing for `IO::FileDescriptor` and introduce a similar `Handle` type.

On Windows, we're actually using C file descriptors (`LibC::Int`) instead of OS handles (`LibC::HANDLE`). It's possible to gain one from the other via `_open_osfhandle` and `_get_osfhandle`. But that's a lot of extra juggling when the APIs acutally all operate on OS handles.

This was also previously mentioned in #14316:

> Ideally we could replace the type of `fd` on Windows entirely, similar to how `Socket#fd` returns a `LibC::HANDLE`, in which case we could also drop `_close` / `_dup2` / `_get_osfhandle` / `_open_osfhandle`. This isn't possible since `IO::FileDescriptor#fd` is public.

I do think it's actually possible to change this. The return type of `IO::FileDescriptor#fd` is `Int` and the documentation states the return type's size is platform-specific. So we should be good. `LibC::HANDLE` is a void pointer. We can return that (maybe casted to int) without breaking any promises.

I'd be fine with breaking the return type on Windows though (i.e. return `LibC::HANDLE` directly), because it's not a stable supported platform yet. And I'd expect minimal impact from such a change.

### PR DESCRIPTION
**Pull request by HertzDevil**
Resolves #14389.

This makes `IO::FileDescriptor#@volatile_fd` an `Atomic(LibC::UIntPtrT)` rather than an `Atomic(LibC::HANDLE)`, to avoid breaking `#fd`'s return type.

### CODE REVIEW

File: src/crystal/system/win32/file_descriptor.cr [hunk 1]
```diff
@@ -9,7 +9,12 @@ module Crystal::System::FileDescriptor
 
   # Platform-specific type to represent a file descriptor handle to the operating
   # system.
-  alias Handle = ::LibC::Int
+  # NOTE: this should really be `LibC::HANDLE`
```
Comment (straight-shoota) [inline src/crystal/system/win32/file_descriptor.cr:12]: **question:** Why is it not then? I'm not sure which it should really be. And I'm wondering if it even matters much. But probably, `UIntPtrT` is just fine.
Anyway, if leaving such a note, we should give a reason.

File: src/crystal/system/win32/file_descriptor.cr [hunk 2]
```diff
@@ -276,8 +268,14 @@ module Crystal::System::FileDescriptor
   end
 
   def self.from_stdio(fd)
```
Comment (straight-shoota) [inline src/crystal/system/win32/file_descriptor.cr:270]: **question:** Is this method even still necessary? AFAIK it was only used for opening the stdio handles.

Comment (HertzDevil) [reply src/crystal/system/win32/file_descriptor.cr:271]: I believe all the console-specific stuffs in the method body are still needed here

Comment (straight-shoota) [reply src/crystal/system/win32/file_descriptor.cr:271]: Ah right. You're only chaning the `ORIGINAL_*` stdio handles. The main handles in `kernel.cr` still use this method.

File: spec/compiler/macro/macro_methods_spec.cr [hunk 3]
```diff
@@ -3692,10 +3692,9 @@ module Crystal
       end
 
       it "reads file (doesn't exist)" do
-        assert_error <<-CRYSTAL,
+        assert_error <<-CRYSTAL
           {{read_file("#{__DIR__}/../data/build_foo")}}
           CRYSTAL
-          "No such file or directory"
```
Comment (straight-shoota) [inline spec/compiler/macro/macro_methods_spec.cr:3698]: **suggestion:** We could test for the common part of the error message, maybe `Error opening file with mode 'r'`?

**Comment by HertzDevil** [comment]
Note to self: `lib/reply/src/term_size.cr` also defines `LibC::STD_OUTPUT_HANDLE` which should be dropped after this PR (this doesn't show up on CI because the interpreter is still disabled)

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**